### PR TITLE
server: pass encryption key with tablet snapshot

### PR DESCRIPTION
**Pull request by tabokie**
Signed-off-by: tabokie <xy.tao@outlook.com>

### What is changed and how it works?

Issue Number: Ref #12842 

What's Changed:

Pass encryption key for each file to snapshot recipient.

```commit-message
```

### Related changes

- https://github.com/pingcap/kvproto/pull/1047

### Check List 

Tests 

- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: src/server/tablet_snap.rs [hunk 1]
```diff
@@ -234,32 +261,39 @@ async fn recv_snap_files(
...
+        let len = data[0] as usize;
+        let file_name = box_try!(std::str::from_utf8(&data[1..len + 1]));
         let p = path.join(file_name);
         let mut f = File::create(&p)?;
-        let mut size = chunk.len() - len - 1;
-        f.write_all(&chunk[len + 1..])?;
+        if chunk.has_encryption_info() {
+            let info = chunk.take_encryption_info();
+            if let Some(mgr) = &key_manager {
+                mgr.insert_file_raw(p.to_str().unwrap(), info)?;
```
Comment (tonyxuqqi) [inline src/server/tablet_snap.rs:279]: Should "fun install_tablet" in components/raftstore-v2/src/operation/ready/snapshot.rs be updated too?
Here the p.to_str() is absolute path and once the whole tablet moved from snapshot folder to tablet folder,  the keys dict should be updated as well.

**Review by tonyxuqqi — COMMENTED** [review]
LGTM except the key handling after applying snapshot seems missed.

File: src/server/tablet_snap.rs (hunk 1 above)
Comment (tabokie) [reply src/server/tablet_snap.rs:279]: Will be handled in https://github.com/tikv/tikv/pull/14095.

File: src/server/tablet_snap.rs [hunk 2]
```diff
@@ -80,21 +84,57 @@ fn is_sst(file_name: &str) -> bool {
...
+        limiter.consume(cost).await;
+        SNAP_LIMIT_TRANSPORT_BYTES_COUNTER_STATIC
+            .send
+            .inc_by(cost as u64);
+        to.clear();
+        to.reserve_exact(size);
+        let mut buf: BorrowedBuf<'_> = to.spare_capacity_mut().into();
+        match &mut self.0 {
+            Either::Left(f) => f.read_buf_exact(buf.unfilled())?,
+            Either::Right(f) => f.read_buf_exact(buf.unfilled())?,
```
Comment (BusyJay) [inline src/server/tablet_snap.rs:131]: Why decrypt the file?

Comment (tabokie) [reply src/server/tablet_snap.rs:131]: Strictly speaking it's possible that different plaintext being encrypted with different keys still output the same results. So checking the plaintext is more correct.

File: src/server/tablet_snap.rs [hunk 3]
```diff
@@ -80,21 +84,57 @@ fn is_sst(file_name: &str) -> bool {
     file_name.ends_with(".sst")
 }
 
-async fn read_to(f: &mut File, to: &mut Vec<u8>, size: usize, limiter: &Limiter) -> Result<()> {
```
Comment (BusyJay) [inline src/server/tablet_snap.rs:83]: You don't need to introduce `WrappedReadableFile`, just changing `&mut File` to `&mut impl Read` should work.

Comment (tabokie) [reply src/server/tablet_snap.rs:83]: To accomplish that, you will need to abstract all code interacting with files into a generic (like `is_sst_match_preview(f: impl Read)`, not just this one function. It is a bigger change than the current way.

Comment (BusyJay) [reply src/server/tablet_snap.rs:83]: Even extracting the single function `read_to` is good for readability. When seeing `WrappedReadableFile` is constructed before sending, it seems the file is decrypted before sending. But turns out it's not as `None` stands for no decryption. If `read_to` doesn't depend on `WrappedReadableFile`, `File` can be used directly, hence no misunderstanding.

**Review by BusyJay — APPROVED** [review]

**Review by SpadeA-Tang — APPROVED** [review]

**Pull request merged**